### PR TITLE
Transformer updates and memory optimizations

### PR DESCRIPTION
**Pull request by tbennun**
- [x] Save memory on backpropagation by not replicating gradients between GradientManager and data_type_optimizer
- [x] Save more memory in FSDP by synchronizing previous outstanding async communication calls and freeing up local gradient contributions
- [x] FSDP: release full weight views after backprop
- [x] Minor tweaks to transformer training script

### CODE REVIEW

File: include/lbann/optimizers/data_type_optimizer.hpp [hunk 1]
```diff
@@ -158,6 +158,7 @@ class data_type_optimizer
   /** @brief Objective function gradient w.r.t. weights (potentially sharded).
    */
   std::unique_ptr<AbsDistMatrixType> m_gradient;
+  friend class optimizer;
```
Comment (benson31) [inline include/lbann/optimizers/data_type_optimizer.hpp:161]: 😬

File: include/lbann/optimizers/optimizer_impl.hpp [hunk 2]
```diff
@@ -27,6 +27,7 @@
 #ifndef LBANN_OPTIMIZERS_OPTIMIZER_IMPL_HPP_INCLUDED
 #define LBANN_OPTIMIZERS_OPTIMIZER_IMPL_HPP_INCLUDED
 
+#include "lbann/optimizers/data_type_optimizer.hpp"
```
Comment (benson31) [inline include/lbann/optimizers/optimizer_impl.hpp:30]: 😬

File: include/lbann/optimizers/optimizer_impl.hpp [hunk 3]
```diff
@@ -218,11 +240,17 @@ optimizer::get_gradient_buffer(TensorDataType& buf_scale,
   // If the manager hasn't been created, let's make it.
   auto mat_info = this->get_matrix_info();
   if (!grad_mgr_ptr) {
+    El::AbstractDistMatrix<TensorDataType>* mat = nullptr;
+    if (dynamic_cast<data_type_optimizer<TensorDataType>*>(this) != nullptr) {
+      mat = dynamic_cast<data_type_optimizer<TensorDataType>*>(this)
+              ->m_gradient.get();
+    }
```
Comment (benson31) [inline include/lbann/optimizers/optimizer_impl.hpp:247]:
```suggestion
    if (auto* dto = dynamic_cast<data_type_optimizer<TensorDataType>*>(this)) {
      mat = dto->m_gradient.get();
    }
```
But also 😬

File: src/callbacks/evaluate_progress.cpp [hunk 4]
```diff
@@ -56,6 +56,7 @@ EvalType evaluate(model& m, const std::string& metric_name, execution_mode mode)
   get_trainer().evaluate(&m, mode);
 
   // Get metric value
+  /*
   bool found_metric = false;
   EvalType metric_value = 0;
```
Comment (benson31) [inline src/callbacks/evaluate_progress.cpp:61]:
```suggestion
  EvalType metric_value = -1.0;
  if (metric_name.size()) {
    bool found_metric = false;
```
Indent the rest as necessary.

File: src/callbacks/evaluate_progress.cpp [hunk 5]
```diff
@@ -73,6 +74,7 @@ EvalType evaluate(model& m, const std::string& metric_name, execution_mode mode)
                 m.get_name(),
                 "\"");
   }
+  */
```
Comment (benson31) [inline src/callbacks/evaluate_progress.cpp:77]: Following on to above suggestion:
```suggestion
  }
```

File: src/callbacks/evaluate_progress.cpp [hunk 6]
```diff
@@ -81,7 +83,8 @@ EvalType evaluate(model& m, const std::string& metric_name, execution_mode mode)
   // Clean up and return metric value
   m.reset_mode(c, original_mode);
   dc.reset_mode(c);
-  return metric_value;
+  return -1.0;
+  // return metric_value;
```
Comment (benson31) [inline src/callbacks/evaluate_progress.cpp:87]: Following on to above suggestions:
```suggestion
  return metric_value;
```

If you don't like these suggestions, please just delete the code and change the return type of the function -- it lives in the anonymous namespace, so I can't imagine that will be too much of a headache.

**Review by benson31 — APPROVED** [review]
LGTM

**Review by benson31 — APPROVED** [review]

**Review by ndryden — APPROVED** [review]
